engine/flink/metrics-deferred-reporter/src/main/java/pl/touk/nussknacker/engine/flink/metrics/MetricRemovalDeferredToNextReportMetricReporter.java

```diff
@@ -0,0 +1,54 @@
+package pl.touk.nussknacker.engine.flink.metrics;
+
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.metrics.Metric;
+import org.apache.flink.metrics.MetricConfig;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.reporter.MetricReporter;
+import org.apache.flink.metrics.reporter.Scheduled;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class MetricRemovalDeferredToNextReportMetricReporter implements MetricReporter, Scheduled {
+
+    private final MetricReporter delegate;
+    private final Set<Tuple3<Metric, String, MetricGroup>> metricsToRemoveAfterNextReport = new HashSet<>();
+
+    public MetricRemovalDeferredToNextReportMetricReporter(MetricReporter delegate) {
+        if (!(delegate instanceof Scheduled)) {
+            throw new IllegalStateException("Only Scheduled reporters are supported");
+        }
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void open(MetricConfig config) {
+        delegate.open(config);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    @Override
+    public void notifyOfAddedMetric(Metric metric, String metricName, MetricGroup group) {
+        delegate.notifyOfAddedMetric(metric, metricName, group);
```

Comment:
This shouldn't matter in practice, but maybe let's remove this added metric from `metricsToRemoveAfterNextReport`, in case that `remove` and `add` are called in quick succession?